R/db.R

```diff
@@ -150,33 +191,146 @@ write_data_safely <- function(dbconn, tblname, df, key) {
+  }
+
+  get_shard_names <- function(df, tblname) {
+    ## Two cases: the shards already exist - or they don't
+    ##
+    ## Fetch existing shards
+    shards <- c()
+    if (DBI::dbExistsTable(dbconn, 'table_shard_map')) {
+      shards <- DBI::dbGetQuery(dbconn, pp("SELECT shard_name FROM table_shard_map WHERE table_name='#{tblname}'"))
+      if (NROW(shards) > 0) {
+        shards <- shards[[1]]
+      }
+    }
+    ## come up with new shards if needed
+    numcols <- NCOL(df)
+    if (numcols == 0) return(NULL)
+    numshards <- ceiling(numcols / MAX_COLUMNS_PER_SHARD)
+    ## All data-containing tables will start with prefix *shard#{n}_*
+    newshards <- paste0("shard", seq(numshards), "_", digest::digest(tblname))
+    if (NROW(shards) > 0) {
+      unique(c(shards, newshards))
+    } else newshards
+  }
+
+  df2shards <- function(dbconn, df, shard_names, key) {
+    ## Here comes the hard part. Sharding strategies!
+    ##
+    ## Here is how we're going to do it.
+    ## We sort the shardnames, to ensure that the first shard is the biggest in size
+    ## This way appending to a shard is trivial: if we have any columns in the
+    ## dataframe that are not yet stored in the cache - just append them to the
+    ## last shard!
+    ## Since we've done the calculation of number of shards beforehand we
+    ## don't even have to worry about creating new shards if something won't fit.
+    ##
+    ## Because it will.
+
+    ## Make sure we don't store `key` in the used_columns! Need it in every dataframe
+    used_columns <- c()
+
+    lapply(sort(shard_names), function (shard, last, key) {
+      ## We need to create a map in the form of
+      ## ```list(df = dataframe, shard_name = shard_names)```, where the dataframe is a subset
+      ## of the original dataframe that contains less columns than
+      ## **MAX_COLUMNS_PER_SHARD**.
+      ## This is what we should do for each shard:
+      ##
+      ## 1. Determine which columns are already being stored in the shard
+      ## 2. Take the subset of the dataframe that has these columns, assign it to a shard
+      ## 3. See which columns are left unsaved, and add those to the last shard
+      if (shard == last) {
+        ## Write out the rest of the dataframe into the last shard
+        list(df = df[setdiff(colnames(df), used_columns)], shard_name = shard)
+      } else {
+        ## If the response is empty, write the first N columns of the dataframe
+        ## Othrwise, only write out those columns that already exist in this shard
+        shard_exists <- DBI::dbExistsTable(dbconn, shard)
+        if (isTRUE(shard_exists)) {
+          one_row <- DBI::dbGetQuery(dbconn, paste("SELECT * FROM ", shard, " LIMIT 1"))
+        } else one_row <- NULL
+        ## Here we abuse the fact that ```NROW(NULL) == 0```
+        if (NROW(one_row) == 0 || NCOL(one_row) == 2) {
+          ## This is very hacky...
+          ## If we see only two columns in a shard, it means that we only stored
```

Comment:
there can never be the case when we meaningfully have 2 columns in a table, because we always duplicate the `key`. 
In fact, we store `num(meaningful_columns) + hashed_key_column + unhashed_key_column_used_for_lookups`
So, if you want to store one piece of meaningful data, you end up storing 3 columns!